components/form/form.js

```diff
@@ -0,0 +1,197 @@
+		this._submit(submitter);
+	}
+
+	_onUnload(e) {
+		if (this.trackChanges && this._dirty) {
+			e.preventDefault();
+			e.returnValue = false;
+		}
+	}
+
+	async _submit(submitter) {
+		const errors = await this.validate();
+		if (errors.length > 0) {
+			return;
+		}
+		let nativeFormData = {};
+		let customFormData = {};
+		const formElements = findFormElements(this);
+		for (const ele of formElements) {
+			const eleData = getFormElementData(ele, submitter);
+			if (isCustomFormElement(ele) || ele === submitter) {
+				customFormData = { ...customFormData, ...eleData };
+			} else {
+				nativeFormData = { ...nativeFormData, ...eleData };
+			}
+		}
+		const formData = { ...nativeFormData, ...customFormData };
+		const event = new CustomEvent('d2l-form-submit', { bubbles: true, cancelable: true, detail: { formData } });
+		if (this.dispatchEvent(event)) {
+			for (const entry of Object.entries(customFormData)) {
+				const input = document.createElement('input');
+				input.type = 'hidden';
+				input.name = entry[0];
+				input.value = entry[1];
+				this._form.appendChild(input);
+			}
+			this._form.submit();
+		}
+	}
+
+	async _validateFormElement(ele, showErrors) {
+		ele.id = ele.id || getUniqueId();
+		if (isCustomFormElement(ele)) {
+			return ele.validate(showErrors);
+		} else if (isNativeFormElement(ele)) {
+			const customs = [...this._validationCustoms].filter(custom => custom.forElement === ele);
+			const results = await Promise.all(customs.map(custom => custom.validate()));
+			const errors = customs.map(custom => custom.failureText).filter((_, i) => !results[i]);
+			if (!ele.checkValidity()) {
+				const validationMessage = localizeFormElement(this.localize.bind(this), ele);
+				errors.unshift(validationMessage);
+			}
+			if (errors.length === 0) {
+				this._validationTooltipHide(ele);
+				ele.setAttribute('aria-invalid', 'false');
+			} else if (showErrors || ele.getAttribute('aria-invalid') === 'true') {
+				this._validationTooltipShow(ele, errors[0]);
+				ele.setAttribute('aria-invalid', 'true');
+			}
+			return errors;
+		}
+		return [];
+	}
+
+	_validationCustomConnected(e) {
+		e.stopPropagation();
+		const custom = e.composedPath()[0];
+		this._validationCustoms.add(custom);
+	}
+
+	_validationCustomDisconnected(e) {
+		e.stopPropagation();
+		const custom = e.composedPath()[0];
+		this._validationCustoms.delete(custom);
+	}
+
+	_validationTooltipHide(ele) {
+		const tooltip = this._tooltips.get(ele);
+		if (tooltip) {
+			this._tooltips.delete(ele);
+			tooltip.remove();
+		}
+	}
+
+	_validationTooltipShow(ele, message) {
+		let tooltip = this._tooltips.get(ele);
+		if (!tooltip) {
+			tooltip = document.createElement('d2l-tooltip');
+			tooltip.for = ele.id;
+			tooltip.align = 'start';
+			tooltip.state = 'error';
+			tooltip.innerHTML = message;
+			ele.parentNode.append(tooltip);
+			this._tooltips.set(ele, tooltip);
+		} else if (tooltip.innerHTML !== message) {
```

Comment:
I'm not sure how we want to handle that. The only reason I use `innerHTML` is because ShadyDOM doesn't play nicely with setting `innerText`.

Reply:
This could also be an HTML-injection vector if message contained malicious HTML. Does this not work?

```javascript
tooltip.appendChild( document.createTextNode( message ) );
```

Reply:
There's also `textContent`.  What difficulty does shadyDOM create for `innerText`?

Reply:
Setting `innerText` adds a node between what would be the `host` and it's `shadowRoot` instead of inside the slot so it completely breaks the component.

Slotted content should be in the `d2l-tooltip-content` `div`:
![Capture](https://user-images.githubusercontent.com/11379933/87986569-4b506600-caab-11ea-87c9-029d14cc28ac.PNG)
![Capture](https://user-images.githubusercontent.com/11379933/87986274-e85ecf00-caaa-11ea-9542-b9b357a955d8.PNG)